packages/terra-navigation-prompt/src/NavigationPromptCheckpoint.jsx

```diff
@@ -0,0 +1,148 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+import PromptRegistrationContext from './PromptRegistrationContext';
+import withPromptRegistration from './_withPromptRegistration';
+import CheckpointNotificationDialog from './_CheckpointNotificationDialog';
+
+const propTypes = {
+  /**
+   * Components to render within the context of the NavigationPromptCheckpoint. Any NavigationPrompts rendered within
+   * these child components will be registered to this NavigationPromptCheckpoint instance.
+   */
+  children: PropTypes.node,
+  /**
+   * A function that will be executed when NavigationPrompts are registered to or deregistered from the NavigationPromptCheckpoint instance.
+   * This can be used to track registered prompts outside of a NavigationPromptCheckpoint and handle prompt resolution directly, if necessary.
+   * The function will be provided with an array of data objects representing the registered NavigationPrompts as the sole argument. An empty
+   * Array will be provided when no prompts are registered.
+   *
+   * Function Example:
+   *
+   * `(arrayOfPrompts) => {
+   *   arrayOfPrompts.forEach((prompt) => {
+   *     console.log(prompt.description);
+   *     console.log(prompt.metaData);
+   *   })
+   *   this.myLocalPromptRegistry = arrayOfPrompts;
+   * }`
+   */
+  onPromptChange: PropTypes.func,
+  /**
+   * @private
+   * An object containing prompt registration APIs provided through the PromptRegistrationContext.
+   */
+  promptRegistration: PropTypes.shape({
+    registerPrompt: PropTypes.func.isRequired,
+    deregisterPrompt: PropTypes.func.isRequired,
+  }).isRequired,
+};
+
+class NavigationPromptCheckpoint extends React.Component {
+  static getPromptArray(prompts) {
+    return Object.keys(prompts).map(id => ({ description: prompts[id][0], metaData: prompts[id][1] }));
+  }
+
+  constructor(props) {
+    super(props);
+
+    this.registerPrompt = this.registerPrompt.bind(this);
+    this.deregisterPrompt = this.deregisterPrompt.bind(this);
+    this.resolvePrompts = this.resolvePrompts.bind(this);
+
+    this.registeredPrompts = {};
+    this.promptProviderValue = {
+      registerPrompt: this.registerPrompt,
+      deregisterPrompt: this.deregisterPrompt,
+    };
+    this.checkpointNotificationDialogRef = React.createRef();
+  }
+
+  componentWillUnmount() {
+    const { onPromptChange } = this.props;
+
+    if (onPromptChange) {
+      /**
+       * The consumer is notified on unmount with an empty set of prompt data to clean up any previously mounted prompts.
+       */
+      onPromptChange([]);
+    }
+  }
+
+  registerPrompt(promptId, promptDescription, metaData) {
+    const { onPromptChange, promptRegistration } = this.props;
+
+    if (!promptId) {
+      return;
+    }
+
+    this.registeredPrompts[promptId] = [promptDescription, metaData];
```

Comment:
Out of curiosity, why store this as an array instead of a hash? I notice above you accessed these like
```
description: prompts[id][0], metaData: prompts[id][1]
```

Reply:
I know that I'm always dealing with these two values, so I figured the looking them up by index would be (ever so slightly) more efficient. Its probably pretty arbitrary with a key space this small.

Reply:
Given that I'm converting ultimately converting to an Object, I may as well make an Object from the get go.

Reply:
68704dd